### PR TITLE
Phase observer (phase 1)

### PR DESCRIPTION
**Pull request by daniela-venuta**

### CODE REVIEW

File: EmpireLegends/GameObserver.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+{
+	_observers->remove(o);
+}
+
+void Observable::Notify()
+{
+	list<Observer*>::iterator i = _observers->begin();
+	for (; i != _observers->end(); i++)
+	{
+		(*i)->Update();
```
Comment (t-andrews) [inline EmpireLegends/GameObserver.cpp:31]: Do you really have to unwrap the pointer here?

File: EmpireLegends/GameObserver.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	}
+}
+
+Observable::Observable()
+{
+	_observers = new list<Observer*>;
```
Comment (t-andrews) [inline EmpireLegends/GameObserver.cpp:37]: just curious. why a `list`  over a `vector`? is it just that you got the code from somewhere? 😆

File: EmpireLegends/GameStart.cpp [hunk 3]
```diff
@@ -92,3 +92,30 @@ void GameStart::detPlayerCount() {
...
+
+void GameStart::startGame()
+{
+	state = "WELCOME TO EMPIRE LEGENDS. THE GAME IS ABOUT TO START.";
+	Notify();
+}
+
+GameStartObserver::GameStartObserver(GameStart* s)
+{
+	subject = s;
```
Comment (t-andrews) [inline EmpireLegends/GameStart.cpp:239]: not sure why the observer would keep an instance of its subject? Isn't the goal of an observer instance to be held by the subject, and not the opposite?

File: EmpireLegends/GameStart.cpp [hunk 4]
```diff
@@ -92,3 +92,30 @@ void GameStart::detPlayerCount() {
...
+	subject->Attach(this);
+}
+
+GameStartObserver::~GameStartObserver()
+{
+	subject->Detach(this);
```
Comment (t-andrews) [inline EmpireLegends/GameStart.cpp:245]: shouldn't detaching observers be the subject's responsibility and not the observer's? That way each subject just detaches their observers all at once when they're destroyed

**Review by t-andrews — CHANGES_REQUESTED** [review]
Let me know if you agree with me about the `subject` stuff. It would apply for all observers

File: EmpireLegends/GameObserver.cpp (hunk 2 above)
Comment (daniela-venuta) [reply EmpireLegends/GameObserver.cpp:37]: lmaoo yeah it was how the prof did it in the lab/ slides lmao

File: EmpireLegends/GameStart.cpp (hunk 3 above)
Comment (daniela-venuta) [reply EmpireLegends/GameStart.cpp:239]: Nope this is how the pattern is implemented :)
https://www.tutorialspoint.com/design_pattern/observer_pattern.htm (also in slides)

File: EmpireLegends/GameObserver.cpp (hunk 1 above)
Comment (daniela-venuta) [reply EmpireLegends/GameObserver.cpp:31]: iterator

**Review by emilie-martin — APPROVED** [review]
looks good to me!

File: EmpireLegends/GameStart.h [hunk 5]
```diff
@@ -1,22 +1,34 @@
...
 	vector<Player*> detPlayerCount();
 	vector<Player*> detPlayerBotCount();
 	vector<Player*> detBotCount();
-	//vector<Player*> getPlayers();
-	
-//private:
-	//int typeMap;
-	//int numOfPlayers;
-	//vector<Player*> players;
-	//vector<Player*> biddingPlayers;
```
Comment (t-andrews) [inline EmpireLegends/GameStart.h:21]: please remove the commented out code

**Review by t-andrews — DISMISSED** [review]
Also, are the filters fixed?

File: EmpireLegends/GameStart.h (hunk 5 above)
Comment (daniela-venuta) [reply EmpireLegends/GameStart.h:21]: done!

**Pull request merged**